std/experimental/color/rgb.d

```diff
@@ -0,0 +1,852 @@
+
+    // create members for some useful information
+    alias ComponentType = ComponentType_;
+    enum string components = components_;
+    enum RGBColorSpace colorSpace = colorSpace_;
+    enum bool linear = linear_;
+    enum bool hasAlpha = CanFind!(components, 'a');
+
+
+    // mixin will emit members for components
+    template Components(string components)
+    {
+        static if(components.length == 0)
+            enum Components = "";
+        else
+            enum Components = (components[0] == 'x' ? "private " : "") ~ ComponentType.stringof ~ ' ' ~ components[0] ~ " = 0;\n" ~ Components!(components[1..$]);
+    }
+    mixin(Components!components);
+
+    // missing components will be fabricated such that code that references them compiles
+    mixin template MissingComponent(char name)
+    {
+        // missing color components will equal the luminance, if it exists. 0 otherwise.
+        static if(!CanFind!(components, 'l'))
+            mixin("enum ComponentType "~name~" = 0;");
+        else
+            mixin("alias "~name~" = l;");
+    }
+    static if(!CanFind!(components, 'r'))
+        mixin MissingComponent!'r';
+    static if(!CanFind!(components, 'g'))
+        mixin MissingComponent!'g';
+    static if(!CanFind!(components, 'b'))
+        mixin MissingComponent!'b';
+    static if(!CanFind!(components, 'l'))
+        @property ComponentType l() const { return toGrayscale!colorSpace(r, g, b); }
+    static if(!CanFind!(components, 'a'))
+        enum ComponentType a = 0;
+
+    // functions that return the color channels as a tuple
+    static if(CanFind!(components, 'l'))
+        @property auto tristimulus() const { auto t = l; return tuple(t, t, t); }
+    else
+        @property auto tristimulus() const { return tuple(r, g, b); }
+    @property auto tristimulusWithAlpha() const { return tuple(tristimulus.expand, a); }
+
+    // RGB/A initialiser
+    this(ComponentType r, ComponentType g, ComponentType b, ComponentType a = 0)
+    {
+        foreach(c; TypeTuple!('r','g','b','a'))
+            mixin(ComponentExpression!("this._ = _;", c, null));
+        static if(CanFind!(components, 'l'))
+            this.l = toGrayscale!colorSpace(r, g, b);
+    }
+
+    // L/A initialiser
+    this(ComponentType l, ComponentType a = 0)
+    {
+        foreach(c; TypeTuple!('l','r','g','b'))
+            mixin(ComponentExpression!("this._ = l;", c, null));
+        static if(CanFind!(components, 'a'))
+            this.a = a;
+    }
+
+    // other color types
+    this(C)(C color) if(isColor!C)
+    {
+        // assignment operator performs format conversion
+        this = color;
+    }
+
+    // hex string initialiser
+    this(C)(const(C)[] hex) if(isSomeChar!C)
+    {
+        fromHex(hex);
+    }
+
+    // read hex strings in the standard forms: (#/$/0x)rgb/argb/rrggbb/aarrggbb
+    void fromHex(S)(const(S)[] hex) if(isSomeChar!S)
+    {
+        static ubyte val(ubyte c)
```

Comment:
`pure nothrow @safe @nogc`?
I have a fear of proliferating the critical error that phobos cross references basically everything everywhere... nothing exists in isolation in phobos.